controllers/topology/reconcile_state.go

```diff
@@ -61,6 +70,93 @@ func (r *ClusterReconciler) reconcileState(ctx context.Context, s *scope.Scope)
+	// creating InfrastructureCluster/ControlPlane objects and updating the Cluster with the
+	// references to above objects.
+	if s.Current.InfrastructureCluster == nil || s.Current.ControlPlane.Object == nil {
+		if err := r.Client.Create(ctx, shim); err != nil {
+			if !apierrors.IsAlreadyExists(err) {
+				return errors.Wrap(err, "failed to create the cluster shim object")
+			}
+			if err := r.Client.Get(ctx, client.ObjectKeyFromObject(shim), shim); err != nil {
+				return errors.Wrapf(err, "failed to read the cluster shim object")
+			}
+		}
+		// Enforce type meta back given that it gets blanked out by Get.
+		shim.Kind = "Secret"
+		shim.APIVersion = corev1.SchemeGroupVersion.String()
+
+		// Add the shim as a temporary owner for the InfrastructureCluster.
+		ownerRefs := s.Desired.InfrastructureCluster.GetOwnerReferences()
+		ownerRefs = append(ownerRefs, *ownerReferenceTo(shim))
+		s.Desired.InfrastructureCluster.SetOwnerReferences(ownerRefs)
+
+		// Add the shim as a temporary owner for the ControlPlane.
+		ownerRefs = s.Desired.ControlPlane.Object.GetOwnerReferences()
+		ownerRefs = append(ownerRefs, *ownerReferenceTo(shim))
+		s.Desired.ControlPlane.Object.SetOwnerReferences(ownerRefs)
+	}
+
+	// If the InfrastructureCluster and the ControlPlane objects has been already created
+	// in previous reconciliation, check if they have already been reconciled by the ClusterController
+	// by verifying the ownerReference for the Cluster is present
+	// In this case, if the ownerReference from the shim object is still there, we can
+	// now get rid of the shim object, garbage collecting all the orphaned objects eventually
+	// created in case of errors in between creating InfrastructureCluster/ControlPlane objects
+	// and updating the cluster with the references to above objects.
+	// If instead the ownerReference from the shim object has been already removed, it means that
+	// the shim objects has been already removed in previous reconciliations.
+	if s.Current.InfrastructureCluster != nil && s.Current.ControlPlane.Object != nil {
+		objectsReconciled := hasOwnerReferenceFrom(s.Current.InfrastructureCluster, s.Current.Cluster) &&
+			hasOwnerReferenceFrom(s.Current.ControlPlane.Object, s.Current.Cluster)
+		objectsStillOwnedByTheShim := hasOwnerReferenceFrom(s.Current.InfrastructureCluster, shim) ||
+			hasOwnerReferenceFrom(s.Current.ControlPlane.Object, shim)
+
+		if objectsReconciled && objectsStillOwnedByTheShim {
+			if err := r.Client.Delete(ctx, shim); err != nil {
+				if !apierrors.IsNotFound(err) {
+					return errors.Wrapf(err, "failed to delete the cluster shim object")
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func clusterShim(c *clusterv1.Cluster) *corev1.Secret {
+	shim := &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: corev1.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-shim", c.Name),
```

Comment:
```suggestion
			Name:      fmt.Sprintf("%s-shim", c.Name),
```
Do we need to worry about cluster names being at the limit and potentially incurring in too long name errors?

Reply:
It'd also be great to prefix instead of suffixing the name with something like `cluster-topology-shim-<sha>`

Reply:
I followed the naming convention that we are already using for cluster certificates, assuming that <cluster-name>-shim  won't create problems given that there is already <cluster-name>-etcd.
I quickly double checked this and it see there is no validation at all in CAPI around name length, and this become a problem when names are used as label filter (this happens for cluster name an machine deployment names at least)
I will open an issue to track this problem

WRT to prefix vs  suffix I'm personally -1 given that there are other secrets generated by CAPI.